### ISSUE TITLE
Investigate the reason of connection timeout with new civetweb

### ISSUE THREAD

**Issue by alkurbatov** [cpp-sc2/cpp-sc2#133]
After update to civetweb v1.16 we experience connection timeout on Windows and game could not be started. Works well at least on OS X.

We need to investigate the reason as it possibly blocks update of the submodule.

**Comment by Nickrader** [comment]
Civetweb:
Results of regression test.  
Good: c1cf4be2d3a1c118c64a815672e215af0772ffbe
Bad: b33c879584e0fce2af0b0be1f6e472e7f7be7ab1

Seems Master branch commit history is changed between some commits in this range.  Rebased?

I was unable to figure out how to get `error_log_file` to output anything or create a file?
https://github.com/civetweb/civetweb/blob/master/docs/UserManual.md#error_log_file

Tried adding option to `sc2_connection.cc` ;  function `StartCivetweb()` 
```	
static const char* LOG_ERROR = "error.log";
	
options[] = {    
    ....
    "error_log_file",
    LOG_ERROR,
    ....
}
```
That did not seem to do anything.

```
https://github.com/civetweb/civetweb/blob/master/docs/UserManual.md
	Embedding with OpenSSL on Windows might fail because of calling convention. 
	To force CivetWeb to use __stdcall convention, add /Gz compilation flag in Visual Studio compiler.
```
Changing calling convention did not fix.

Common point of divergence seems to happens sometime after `setsockopt` gets called in `civetweb.c` , however this enters loop shortly after and in good commit, after 4-8? passes, connects to game.  Bad commit times out.  I get lost here.
```
    good commit: pollres == 1 (eventually)
    bad: pollres != 1 (ever)
```

The answer may lie in some error logs, I'll continue to play around trying to access, but this may be as far as I can follow at the moment.  A lot of stuff happens in `civetweb.c` that I don't follow at the moment, my first time looking into a web server.

**Comment by alkurbatov** [comment]
>Seems Master branch commit history is changed between some commits in this range. Rebased?

If by master you mean my fork of civetweb, it was rebased with changes in history (bad approach although simple).

I would suggest to try the latest version of civetweb. It could contains fixes of the problem.

Another option here is to try civetweb without these 2 patches from Blizzard as they could brake something.

**Comment by Nickrader** [comment]
I forked directly from Civetweb then rebased Blizz changes on that. 

The change in history around the problem commit just stood out to me, probably because I spent much time going back and forth between good/bad threshold (double checking my results).  They could have done it other places too, and I just didn't notice.  It was noteworthy, but I don't think it was the problem.

Trying the latest with/without sc2 changes still times out.
Latest Civetweb does work under Windows when I build it alone and test `civetweb.exe`.

I'm assuming some mismatch between the change they made on [b33c879584e0fce2af0b0be1f6e472e7f7be7ab1](https://github.com/civetweb/civetweb/commit/b33c879584e0fce2af0b0be1f6e472e7f7be7ab1) and the way our handler works?  

But I really don't know.  Getting it drilled down to the breaking commit may be as far I get here.

**Comment by Nickrader** [comment]
https://github.com/civetweb/civetweb/commit/b33c879584e0fce2af0b0be1f6e472e7f7be7ab1 (the breaking commit)
Lines 9571 - 9575 correspond to being in `getsockopt` function.  
Which is the same function we added sc2 changes.  I think that is why I assume it has something to do with this.

**Comment by alkurbatov** [comment]
Oh, that is close but the actual problem looks bigger.
The patch with no delay is applied because of this issue https://github.com/civetweb/civetweb/issues/434
Seems that now civetweb uses new interface so we should reimplement the socket routine to make it work (probably without a patch).

**Comment by alkurbatov** [comment]
After taking a closer look, I think there are several improvements we should do to get more info.
The main use case is sc2:: Connection part of the code as server used only to work as a proxy.

1. Try to either enable logging inside civetweb, or provide proper callback functions and log on our side, see:
https://github.com/civetweb/civetweb/blob/master/docs/api/mg_callbacks.md
2. Print ebuff on connection failure. No idea why it is ignored in the current version.

Speaking about the no-delay patch itself:
1. Check the new civetweb API and options to pass no delay.
2. Probably we should reimplement the patch as currently it breaks the client logic by some reason.

**Comment by Nickrader** [comment]
If I try to follow along in the debugger, we call `mg_poll()` and `poll()` in function `pull_inner()`.  There I get lost in macros and a do...while loop.  In the regressing commit, they modify the array size of `struct mg_pollfd`.

Point of divergence, https://github.com/Nickrader/civetweb/blob/9da7ba333fef54db11cadab2a92ef78575672d1a/src/civetweb.c#L6460 .
Here in the working versions, the response is received [nread is given a number which eventually makes it way to `get_message()` ...`conn->request_len`]. 
Call Stack:
	all_tests.exe!pull_inner(_iobuf * fp, mg_connection * conn, char * buf, int len, double timeout) Line 6373	C
 	all_tests.exe!read_message(_iobuf * fp, mg_connection * conn, char * buf, int bufsiz, int * nread) Line 11104	C
 	all_tests.exe!get_message(mg_connection * conn, char * ebuf, unsigned __int64 ebuf_len, int * err) Line 18542	C
 	all_tests.exe!get_response(mg_connection * conn, char * ebuf, unsigned __int64 ebuf_len, int * err) Line 18693	C
 	all_tests.exe!mg_connect_websocket_client_impl(const mg_client_options * client_options, int use_ssl, char * error_buffer, unsigned __int64 error_buffer_size, const char * path, const char * origin, const char * extensions, int(*)(mg_connection *, int, char *, unsigned __int64, void *) data_func, void(*)(const mg_connection *, void *) close_func, void * user_data) Line 19070	C
 	all_tests.exe!mg_connect_websocket_client(const char * host, int port, int use_ssl, char * error_buffer, unsigned __int64 error_buffer_size, const char * path, const char * origin, int(*)(mg_connection *, int, char *, unsigned __int64, void *) data_func, void(*)(const mg_connection *, void *) close_func, void * user_data) Line 19187	C
 	all_tests.exe!sc2::Connection::Connect(const std::string & address, int port, bool verbose) Line 119	C++
 	all_tests.exe!sc2::ProtoInterface::ConnectToGame(const std::string & address, int port, int timeout_ms) Line 71	C++
 	all_tests.exe!sc2::ControlImp::Connect(const std::string & address, int port, int timeout_ms) Line 1478	C++
 	all_tests.exe!sc2::AttachClients(sc2::ProcessSettings & process_settings, std::vector<sc2::Client *,std::allocator<sc2::Client *>> clients) Line 97	C++
 	all_tests.exe!sc2::LaunchProcesses(sc2::ProcessSettings & process_settings, std::vector<sc2::Client *,std::allocator<sc2::Client *>> clients, int window_width, int window_height, int window_start_x, int window_start_y) Line 122	C++
 	all_tests.exe!sc2::Coordinator::LaunchStarcraft() Line 736	C++
 	all_tests.exe!sc2::TestAbilityRemap(int argc, char * * argv) Line 224	C++
 	all_tests.exe!main(int argc, char * * argv) Line 39	C++
 	all_tests.exe!invoke_main() Line 79	C++
 	all_tests.exe!__scrt_common_main_seh() Line 288	C++
 	all_tests.exe!__scrt_common_main() Line 331	C++
 	all_tests.exe!mainCRTStartup(void * __formal) Line 17	C++
 	kernel32.dll!00007fffc2017344()	Unknown
 	ntdll.dll!00007fffc37626b1()	Unknown

TL;DR:  I suspect changing `struct mg_pollfd` did something I do not understand.  "It could be a me problem", I tell myself until another machine verifies it.  Many calls happen to `mg_poll()`, information overload stops me going further.
It's non-breaking since we can use earlier commits, at this point I guess it is just annyoing and taunting us ;)

**Comment by alkurbatov** [comment]
I think that you are exploring wrong direction. mg_poll and any other poll polls the socket used by server (or client) to notify running program when new network data was received. It could be broken in some way but what I think s broken is our custom patch which sets no delay to the client socket.
So I would try to narrow down the issue:
1. First of all does it work without our patch on the "bad" commit?
2. Does it work without our patch on the last commit?
3. If it does in both cases, we should rework the patch in some way. The simplest decision is to reuse civetweb's function to which enables no delay on socket. It has some checks that could be important.

To be honest you are right: this is not a big issue and we can happily live with it for the next year. The only problem is changes in compilers (especially VS) which could cause compilation errors in civetweb. As a result, lack updates of civetweb forces us to do more patches (like it was done before) or reject newer VS/g++/clang (unfortunately not possible).

Thus I believe we should work on the issue from time to time "without fanaticism" as someday the real reason could be found.

**Comment by LairdWT** [comment]
in the thirdparty/civetweb/civetweb.cmake changing the value of the GIT_TAG on line 38 to v1.15 resolved the connection timeout issue for me
![{4BBC6157-2484-4A61-8103-24328775EFA7}](https://github.com/user-attachments/assets/9d910fc5-b9e6-48b7-8aeb-ae715818cbf0)

**Comment by alkurbatov** [comment]
Hi @LairdWT 
Thank you for the info!

@Nickrader could you please check that v1.15 works for you?

**Comment by Nickrader** [comment]
Not near my Windows machine for some time.  I assume it works fine though.

Sept 1, 2021 ; Tag 1.15

Jul 17, 2022 ; [Commit in use prior ](https://github.com/civetweb/civetweb/commit/1fb204ecc630515d53291f58955c799785cb90c7)

May 14, 2023 ; [Breaking Commit](https://github.com/civetweb/civetweb/commit/b33c879584e0fce2af0b0be1f6e472e7f7be7ab1)

**Comment by alkurbatov** [comment]
Ok, I  downgraded civetweb. Let's keep this issue open for further investigation as we have to update civetweb sooner or later.

**Comment by angrylemur** [comment]
Today I tried building and running blank-bot (using cpp-sc2 v2.0.0 tag) in Windows 10, Visual Studio 2022, and I get errors, with the game failing to launch (black SC2 window contents):

Launched SC2 (C:\Program Files (x86)\StarCraft II\Versions\Base93333\SC2_x64.exe), PID: 53368
Connecting to 127.0.0.1:8167...
Failed to start civetweb server: Failed to setup server ports
Failed to start civetweb server: Failed to setup server ports
...

After asking the Starcraft 2 AI Arena discord, I tried downgrading my fetch of cpp-sc2 to f4e8761ee6346c95aa0253d458df241631a4e778 and that ended up working... the game launched and ran correctly.

**Comment by alkurbatov** [comment]
Must work after #161
Requires confirmation.
